File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        using namespace std;
        unordered_set<ListNode*> nodes;
        ListNode* curr = head;
        while (curr->next != nullptr) {
            if (nodes.count(curr)) return true;
            nodes.insert(curr);
            curr = curr->next;
        }
        return false;
    }
};

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode counter = head;
        int count = 0;
        while (counter != null) {
            count++;
            counter = counter.next;
        }
        // remove is now the index of the node to remove
        int remove = count - n;
        if (remove == 0) return head.next;
        counter = head;
        ListNode prev = null;
        while (remove > 0) {
            prev = counter;
            counter = counter.next;
            remove--;
        }
        prev.next = counter.next;
        return head;
    }
}
